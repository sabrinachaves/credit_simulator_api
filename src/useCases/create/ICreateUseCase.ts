import { ICreditSimulation } from '@domain/entities/CreditSimulation';

export interface ICreateCreditInput {
  id?: string;
  amount: number;
  paymentTermInMonths: number;
  birthDate: Date;
}

export default interface ICreateUseCase {
  handle(data: ICreateCreditInput): Promise<ICreditSimulation>;
}
