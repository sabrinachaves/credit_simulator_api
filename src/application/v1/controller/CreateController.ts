import { buildErrorInfo } from '@infrastructure/parser/ErrorInfo';
import ICreateUseCase, { ICreateCreditInput } from '@useCases/create/ICreateUseCase';
import { OK } from 'http-status';
import { Request, Response } from 'express';

export default class CreateController {
  constructor(private createUseCase: ICreateUseCase) {}

  public execute = async (request: Request, response: Response) => {
    try {
      const { amount, paymentTermInMonths, birthDate }: ICreateCreditInput = request.body;
      const credit = await this.createUseCase.handle({ amount, paymentTermInMonths, birthDate });

      return response.status(OK).json(credit);
    } catch (err: any) {
      const errorInfo = buildErrorInfo(err);

      console.error('failed to create a credit simulation', errorInfo);
      return response.status(errorInfo.code).json({ error: errorInfo.errorMessage });
    }
  };
}
