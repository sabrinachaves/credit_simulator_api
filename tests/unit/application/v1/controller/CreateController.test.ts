import { Request } from 'express';
import CreateController from '../../../../../src/application/v1/controller/CreateController';
import CreateUseCaseMock from '../../../../mocks/CreateUseCaseMock';
import { faker } from '@faker-js/faker';
import { INTERNAL_SERVER_ERROR, OK, UNPROCESSABLE_ENTITY } from 'http-status';
import UnprocessableEntityError from '../../../../../src/domain/exceptions/UnprocessableEntityError';
import InternalServerError from '../../../../../src/domain/exceptions/InternalServerError';
import ICreateUseCase from '../../../../../src/useCases/create/ICreateUseCase';

describe('CreateController', () => {
  const createUseCase = new CreateUseCaseMock();

  it('should handle the request #unit', async () => {
    const instance = new CreateController(createUseCase);
    const req = {
      body: {
        birthDate: new Date('2005-05-01'),
        amount: 15000,
        paymentTermInMonths: 15,
      },
    } as Request;
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: jest.fn(),
      })),
    };

    createUseCase.handle = jest.fn().mockResolvedValue({
      id: faker.datatype.uuid(),
      amount: 15000,
      paymentTerm: 15,
      birthDate: new Date('2005-05-01'),
      amountToBePaid: 15504.9,
      installmentsValue: 1033.66,
      amountPaidInInterest: 504.9,
      interestRate: 5,
    });

    await instance.execute(req, res);

    expect(createUseCase.handle).toHaveBeenCalledWith({
      birthDate: new Date('2005-05-01'),
      amount: 15000,
      paymentTermInMonths: 15,
    });
    expect(res.status).toHaveBeenCalledWith(OK);
  });

  it('should throw an error when age is less than 18 #unit', async () => {
    const instance = new CreateController(createUseCase);
    const req = {
      body: {
        birthDate: new Date('2010-05-01'),
        amount: 15000,
        paymentTermInMonths: 15,
      },
    } as Request;
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: jest.fn(),
      })),
    };

    createUseCase.handle = jest
      .fn()
      .mockRejectedValueOnce(new UnprocessableEntityError('Age can not be less than 18 years old'));

    await instance.execute(req, res);

    expect(createUseCase.handle).toHaveBeenCalledWith({
      birthDate: new Date('2010-05-01'),
      amount: 15000,
      paymentTermInMonths: 15,
    });
    expect(res.status).toHaveBeenCalledWith(UNPROCESSABLE_ENTITY);
  });

  it('should throw an error when useCase throws #unit', async () => {
    const instance = new CreateController(createUseCase);
    const req = {
      body: {
        birthDate: new Date('2010-05-01'),
        amount: 15000,
        paymentTermInMonths: 15,
      },
    } as Request;
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: jest.fn(),
      })),
    };

    createUseCase.handle = jest.fn().mockRejectedValueOnce(new InternalServerError('Internal Server Error'));

    await instance.execute(req, res);

    expect(createUseCase.handle).toHaveBeenCalledWith({
      birthDate: new Date('2010-05-01'),
      amount: 15000,
      paymentTermInMonths: 15,
    });
    expect(res.status).toHaveBeenCalledWith(INTERNAL_SERVER_ERROR);
  });

  it('should match the snapshot #unit', () => {
    const instance = new CreateController({} as ICreateUseCase);
    expect(instance).toMatchSnapshot();
  });
});
