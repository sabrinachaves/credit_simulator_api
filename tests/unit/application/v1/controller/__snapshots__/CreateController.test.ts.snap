// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateController should match the snapshot #unit 1`] = `
CreateController {
  "createUseCase": Object {},
  "execute": [Function],
}
`;
